[Review] feat(sec): Create certificate in C using OpenSSL/LibreSSL

ccvca opened the pull request:
Create a self-signed encryption certificate in C (only implemented with OpenSSL/LibreSSL).

For demonstration purpose, `UA_CreateCertificate` is implemented as a fallback in `server_encryption.c`, if no files are passed using the command line.

Resolves: #2816

GoetzGoerisch commented:
Missing Mbed TLS support could be implemented after #4372 is solved.

ccvca commented:
The  "AppVeyor"-Status looks goot, but Github does not show the correct result.

jpfr commented on include/open62541/plugin/securitypolicy.h:
```diff
@@ -320,6 +321,31 @@ struct UA_SecurityPolicy {
     void (*clear)(UA_SecurityPolicy *policy);
 };
 
+#ifdef UA_ENABLE_ENCRYPTION
```
Move this definition to the plugins/include folder.

jpfr commented on include/open62541/plugin/securitypolicy.h:
```diff
@@ -320,6 +321,31 @@ struct UA_SecurityPolicy {
+};
+
+/**
+ * (Experimental) Create a certificate
```
Mention that this is a self-signed certificate.

jpfr commented on include/open62541/plugin/securitypolicy.h:
```diff
@@ -320,6 +321,31 @@ struct UA_SecurityPolicy {
+ *                  e.g. ["DNS:localhost", "URI:urn:open62541.server.application"]
+ */
+UA_StatusCode UA_EXPORT
+UA_CreateCertificate(const UA_Logger *logger,
```
Keep consistent with the style of the overall SDK.
Array length are size_t.
The name of the array length is "arrayNameSize".
With a `Size` postfix.

jpfr commented on plugins/crypto/mbedtls/ua_mbedtls_create_certificate.c:
```diff
@@ -0,0 +1,30 @@
+                    enum UA_CertificateFormat certFormat) {
+    if(!outPKey || !outCert || !logger || !subjectAltName
+        || !subject || lenSubjectAltName == 0 || lenSubject == 0) {
+        return UA_STATUSCODE_BADINVALIDARGUMENT;
```
If it builds, then it should work.
I prefer to have this function undefined for mbedtls (linker error for a missing function) than to let it fail only at runtime.
That reduces the potential surprise at runtime.

jpfr commented on plugins/crypto/openssl/ua_openssl_create_certificate.c:
```diff
@@ -0,0 +1,366 @@
+ * Join an array of UA_String to a single NULL-Terminated UA_String
+ * separated by character sep
+ */
+static UA_StatusCode UA_String_join_nullterm(
```
"nullterm" is specialized nomenclature not verybody knows.
Call this something easy to understand. Like join_string_with_sep or similar.

Format this definition according to CONTRIBUTING.md.
See the comments on the public header for the representation of arrays.

jpfr commented on plugins/crypto/openssl/ua_openssl_create_certificate.c:
```diff
@@ -0,0 +1,366 @@
+    char sep,
+    UA_String *out) {
+    if(!out)
+    {
```
We don't need all these braces/newlines here.

jpfr commented on plugins/crypto/openssl/ua_openssl_create_certificate.c:
```diff
@@ -0,0 +1,366 @@
+    }
+    UA_String_clear(out);
+    size_t totalSize = lenStrings;
+    size_t pos = 0;
```
Move the definition of pos below.
Where it is first used.

jpfr commented on plugins/crypto/openssl/ua_openssl_create_certificate.c:
```diff
@@ -0,0 +1,366 @@
+        totalSize += strings[iStr].length;
+    }
+
+    out->length = totalSize;
```
UA_ByteString_allocBuffer is shorter / cleaner.

jpfr commented on plugins/crypto/openssl/ua_openssl_create_certificate.c:
```diff
@@ -0,0 +1,366 @@
+
+    UA_Byte byteNeedl = (UA_Byte)needl;
+
+    for(UA_Int32 i = 0; (size_t)i < pUaStr->length; ++i) {
```
indices are size_t.

jpfr commented on plugins/crypto/openssl/ua_openssl_create_certificate.c:
```diff
@@ -0,0 +1,366 @@
+            errRet = UA_STATUSCODE_BADINTERNALERROR;
+            goto cleanup;
+        }
+        memcpy(field, subject[iSubject].data, (size_t) sep);
```
You can just replace the '=' with a '\0' in the original string.
Than you don't have to copy to a temporary variable and can just forward the original string with the name.
The replacement can be rolled back later. So that the original string is left identical when returning.

ccvca replied:
This causes a SegFault, when using `UA_STRING_STATIC`.

jpfr commented on plugins/crypto/openssl/ua_openssl_create_certificate.c:
```diff
@@ -0,0 +1,366 @@
+            break;
+        }
+        default: {
+            UA_LOG_ERROR(logger, UA_LOGCATEGORY_SECURECHANNEL,
```
set the return code to something bad.
this could be checked early on in the function.

jpfr reviewed (commented):
Looks great!
A couple of comments in the technical details.
Can be merged then.

ccvca commented:
I've resolved all comments except one, which causes a SegFault.

jpfr commented:
So why don’t the unit tests fail?

try to pinpoint the problem with Valgrind.
Often times it becomes quite obvious when doing that.

ccvca commented:
I've investigate the problem a little bit further using this short snippet:
```c
#include <open62541/types.h>

void func(UA_String testStr) {
    UA_Byte orig = testStr.data[1];
    testStr.data[1] = 'D';
    testStr.data[1] = orig;
}

int main(int argc, char* argv[]) {
    UA_String testStr = UA_STRING_STATIC("ABC");
    func(testStr);
    return 0;
}
```

Running this application without an debugger is fine (at least no crash). But gdb doesn't like it:

```txt
GNU gdb (GDB) 8.1
Copyright (C) 2018 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.  Type "show copying"
and "show warranty" for details.
This GDB was configured as "i686-w64-mingw32".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
<http://www.gnu.org/software/gdb/documentation/>.
For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from D:\Programmieren\Cpp\Open62541_Tests\Test_StaticSEGFAULT\build\Test_StaticMem_SegFault.exe...done.
(gdb) r
Starting program: D:\Programmieren\Cpp\Open62541_Tests\Test_StaticSEGFAULT\build\Test_StaticMem_SegFault.exe 
[New Thread 26552.0x461c]
[New Thread 26552.0x5ae4]
[New Thread 26552.0x1fc]

Thread 1 received signal SIGSEGV, Segmentation fault.
func (testStr=...) at D:\Programmieren\Cpp\Open62541_Tests\Test_StaticSEGFAULT\Test_StaticMem_SegFault.cpp:6
6           testStr.data[1] = 'D';
(gdb) c
Continuing.

Thread 1 received signal SIGSEGV, Segmentation fault.
func (testStr=...) at D:\Programmieren\Cpp\Open62541_Tests\Test_StaticSEGFAULT\Test_StaticMem_SegFault.cpp:6
6           testStr.data[1] = 'D';
(gdb) c
Continuing.
[Thread 26552.0x1fc exited with code 3221225477]
[Thread 26552.0x5ae4 exited with code 3221225477]
[Inferior 1 (process 26552) exited with code 030000000005]
(gdb) q
```

DrMemory also does not totally agree on this:
```txt
~~Dr.M~~ Dr. Memory version 2.3.18722
~~Dr.M~~ Running "Test_StaticMem_SegFault.exe"
~~Dr.M~~ Using system call file c:\Portable\DrMemory-Windows-2.3.18722\drmemory\logs\symcache\syscalls_wow64.txt
~~Dr.M~~ 
~~Dr.M~~ Error #1: WARNING: writing to readonly memory 0x00404045-0x00404046
~~Dr.M~~ # 0 func               [D:/Programmieren/Cpp/Open62541_Tests/Test_StaticSEGFAULT/Test_StaticMem_SegFault.cpp:6]
~~Dr.M~~ # 1 main               [D:/Programmieren/Cpp/Open62541_Tests/Test_StaticSEGFAULT/Test_StaticMem_SegFault.cpp:12]
~~Dr.M~~ Note: @0:00:00.365 in thread 12276
~~Dr.M~~ Note: instruction: mov    $0x44 -> (%eax)
```

So I don't think the proposed optimisation is a good idea.

jpfr commented on plugins/crypto/openssl/ua_openssl_create_certificate.c:
```diff
@@ -0,0 +1,365 @@
+ * Join an array of UA_String to a single NULL-Terminated UA_String
+ * separated by character sep
+ */
+static UA_StatusCode join_string_with_sep(
```
Fix the alignment.
Use UA_String *strings instead of UA_String strings[].
Do this consistent everywhere.

jpfr commented on plugins/include/open62541/plugin/create_certificate.h:
```diff
@@ -0,0 +1,39 @@
+/**
+ * (Experimental) Create a self-signed certificate
+ *
+ * It is recommended to store the generated certificate in a non volatile memory and
```
volatile has a specific meaning in C.
If you mean "store on disk for reuse", just say so.

jpfr commented on plugins/include/open62541/plugin/create_certificate.h:
```diff
@@ -0,0 +1,39 @@
+ *                  e.g. ["DNS:localhost", "URI:urn:open62541.server.application"]
+ */
+UA_StatusCode UA_EXPORT
+UA_CreateCertificate(const UA_Logger *logger,
```
Please use pointers instead of array syntax to be consistent with the rest of the SDK.

Is it possible to generalize this function for different key lengths?
That would be great so we don't have to change the API later when a different key length is requested.

ccvca replied:
Parameter is added.
In case a `0` is provided, the largest possible key is generated (4096Bits in the current OPC UA Specification).

jpfr reviewed (commented):
Ok. Then leave out the UA_STRING_STATIC part.
This is probably because the string memory is marked as const.
So the compiler puts it into a location that is write-protected.

The const-ness of the string however is not visible in the code part where UA_STRING_STATIC could have been used. By the way, UA_STRING is better than UA_STRING_STATIC. Because UA_STRING uses strlen to find the \0 for the length. 
UA_STRING_STATIC uses sizeof() for the length and only works when you put in a string literal.

Otherwise, I have a few minor comments and one request to generalize the key lenght in the API. Let's bring this in soon!

ccvca commented:
I do not see the docker build, but it also failes on the current master.
PS: ghcr.io (github container registry) may be used to provide docker images with github actions instead of an external ci.

jpfr commented:
Hey @ccvca ,

I just added a commit on top of your branch (I can do that for PR to this repo).
A few cosmetic cleanups and a few resolved bugs that came up from valgrind.

Also, valgrind-testing in the CI is now also enabled for OpenSSL.

Please have a look at the cleanup-commit and fold the changes into your history.

Can be merged when you are done.

This is really cool!

nokem61 commented on plugins/crypto/openssl/ua_openssl_create_certificate.c:
```diff
@@ -0,0 +1,368 @@
+        goto cleanup;
+    }
+
+    if(X509_gmtime_adj(X509_get_notAfter(x509), (UA_Int64) 60 * 60 * 24 * 356) == NULL) {
```
356??

ccvca replied:
Changed to 365 :)

ccvca commented:
> 
> 
> Hey @ccvca ,
> 
> I just added a commit on top of your branch (I can do that for PR to this repo).
> A few cosmetic cleanups and a few resolved bugs that came up from valgrind.
> 
> Also, valgrind-testing in the CI is now also enabled for OpenSSL.
> 
> Please have a look at the cleanup-commit and fold the changes into your history.
> 
> Can be merged when you are done.
> 
> This is really cool!

Thank you.

I've cleaned up the commits.

jpfr commented:
Nice. Merged.
